### ISSUE TITLE
Improve query interpretation format

### ISSUE THREAD

**Issue by plajjan** [SpriteLink/NIPAP#683]
The current interpretation format is not flexible and powerful enough to describe more complex queries. It makes certain assumptions but with the introduction of a more complex query language, those assumptions will fail.

This is closely related to #531 but I'm splitting it out as this can be implemented on it's own and I think it is easier for the everyone with as small changes as possible.

The interpretation format needs to be changed as it is not flexible enough to express more complex queries. I would suggest that the interpretation should be baked in together with the dictSQL struct - the two structures are so similar so having them separate is just a waste. In the combined format there would be certain keys that pertain to the core functionality of dictSQL (ie forming a query that can be translated into SQL) and we would have other keys that help with printing an interpretation.

For example, to match within a prefix we have the following dictSQL:
```
{
   'operator': 'contained_within_equals',
   'val1': 'prefix',
   'val2': '1.0.0.0/24'
}
```

Including the interpretation would make it look like:
```
{
   'operator': 'contained_within_equals',
   'val1': 'prefix',
   'val2': '1.0.0.0/24',
   'interpretation': {
      'interpretation': 'IPv4 prefix',
      'string': '1.0.0.3/24',
      'attribute': 'prefix',
      'text': 'Prefix must be contained within 1.0.0.0/24, which is the base prefix of 1.0.0.3/24.',
      'operator': 'contained_within_equals',
      'strict_prefix': '1.0.0.0/24'
   }
}
```

All the keys are copied from the current interpretation format, except for "text" which I've added. In our current implementation we let the client (web or CLI) parse the interpretation dict and format a suitable text message but I'm leaning towards doing this on the server side to avoid writing as much code in both clients. The reason for the current implementation is that a) at first it only existed in the web UI and b) I wanted to be able to put stuff in bold or apply other text formatting. Doing that server side doesn't make sense as the CLI wouldn't want HTML tags for text formatting...

My suggestion going forward is that we simply skip this formatting, I think the messages are clear enough without putting the prefix in bold or similar and that we move to generating the message on the server side for simplicity.

The first question is thus, is it OK to merge these interpretation-keys into the dictSQL and is the above a suitable way?

Second question pertains to where we generate the actual text messages. Client side like today or should we move to server side?

I would also like to show a more complex query, in this case for mathing a tag. Searching for '#foo' will actually expand not only to match on tags=#foo but also on the inherited_tags attribute;
```
{
   'operator': 'or',
   'val2': {
      'operator': 'equals_any',
      'val2': 'foo',
      'val1': 'inherited_tags'
   },
   'val1': {
      'operator': 'equals_any',
      'val2': 'foo',
      'val1': 'tags'
   }
}
```

With the interpretation it looks like this:
```
{
   'operator': 'or',
   'val2': {
      'operator': 'equals_any',
      'val2': 'foo',
      'val1': 'inherited_tags'
   },
   'val1': {
      'operator': 'equals_any',
      'val2': 'foo',
      'val1': 'tags'
   }
   'interpretation': {
      'operator': 'equals_any',
      'attribute': 'tag',
      'interpretation': '(inherited) tag)',
      'string': '#foo',
      'text': 'bar foo'
   }
}
```
So instead of having an interperation for attribute "tag", one for attribute "inherited_tag" and yet another one for the "OR" operator, we just have one. On the client we just print interpretation messages where we find them so while it isn't actually a 100% accurate representation of the query it does make it a lot easier to read. CLI output would be:
```
kll@lingloi320 ~/kod/NIPAP/nipap-cli $ ./nipap address list --show-interpretation '#foo'
Searching for prefixes in any VRF...
Query interpretation:
 - #foo: (inherited) tag) must contain #foo
    The tag(s) or inherited tag(s) must contain #foo
VRF            Prefix                        #  Node                Order          Customer       Description
=============================================================================================================
-              1.0.0.128/25                A #1 None                None           None           test
```

**Comment by plajjan** [comment]
This is the kind of HTML formatting I'm talking about:
![nipap-interpretation-html-bold](https://cloud.githubusercontent.com/assets/841628/5697226/471a8b06-99ec-11e4-9b57-bb6cafbe3ecc.png)
with a server side formatting we would instead get

**(test|link)**: text matching (test|link)

the first **(test|link)** part is still printed by the client and we keep that in the interpretation dict for easy reference and what not while the rest of the line comes from the server.

**Comment by plajjan** [comment]
@garberg feedback plxz =)

**Comment by garberg** [comment]
Adding the query string interpretation to a dictSQL seems as a great idea, I'm all for it.

Generating text messages at server-side also seems as a good idea in general, but we should really make sure to convey all information needed to generate the message in a structured format as well, for example to generate text messages in different languages client-side. We had an offline discussion regarding this and decided to wait with server-side text generation as the client-side is already implemented.

**Comment by plajjan** [comment]
Fixed through merge of #720. Closing.
